### PR TITLE
Remove default parameters from detail headers in include

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Contributes to #9854. None of these changes should affect users, nor do they impose a particularly onerous burden on libcudf developers (just some extra passing through `mr` or `cudf::get_default_stream()`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/copying/scatter.cu
Comment (davidwendt) [inline cpp/src/copying/scatter.cu:396]:
```suggestion
    detail::apply_boolean_mask(table_view{{indices->view()}}, boolean_mask, stream, rmm::mr::get_current_device_resource());
```
This appears to be a temporary and so should not use the passed in `mr`

File: cpp/src/sort/segmented_sort_impl.cuh
Comment (davidwendt) [inline cpp/src/sort/segmented_sort_impl.cuh:77]: This temporary should not be created with the passed in `mr`
https://github.com/rapidsai/cudf/blob/branch-23.04/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#temporary-memory

File: cpp/src/stream_compaction/distinct.cu
Comment (davidwendt) [inline cpp/src/stream_compaction/distinct.cu:149]: The `gather_map` is not returned to the caller and so should use `rmm::mr::get_current_device_resource()` instead of `mr`

File: cpp/src/stream_compaction/stable_distinct.cu
Comment (davidwendt) [inline cpp/src/stream_compaction/stable_distinct.cu:42]:
```suggestion
    get_distinct_indices(input.select(keys), keep, nulls_equal, nans_equal, stream, rmm::mr::get_current_device_resource());
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Looks good. 
Some issues with using `mr` variable for temporary objects.

File: cpp/src/copying/scatter.cu
Comment (vyasr) [reply cpp/src/copying/scatter.cu:396]: Ah thanks I wasn't careful enough in checking on these. Will fix.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/sort/segmented_sort_impl.cuh
Comment (vyasr) [reply cpp/src/sort/segmented_sort_impl.cuh:73]: Ah good call I removed its usage for temporary allocations but didn't realize I'd gotten rid of all uses.

File: cpp/src/stream_compaction/distinct_count.cu
Comment (davidwendt) [inline cpp/src/stream_compaction/distinct_count.cu:17]:
```suggestion
#include <cudf/utilities/default_stream.hpp>
```
This should also be moved below the other 2 includes.

File: cpp/src/stream_compaction/unique_count.cu
Comment (davidwendt) [inline cpp/src/stream_compaction/unique_count.cu:17]:
```suggestion
#include <cudf/utilities/default_stream.hpp>
```
Same here

File: cpp/src/sort/segmented_sort_impl.cuh
Comment (davidwendt) [inline cpp/src/sort/segmented_sort_impl.cuh:27]: Move this include up with the other `include <rmm` line

**Review by davidwendt — CHANGES_REQUESTED** [review]
Looks good. Just some `#include` fixes.

File: cpp/src/stream_compaction/unique_count.cu
Comment (vyasr) [reply cpp/src/stream_compaction/unique_count.cu:17]: Did we add a pre-commit hook or something that added these? I don't even recall manually adding them... @bdice does this sound familiar to you?

Comment (bdice) [reply cpp/src/stream_compaction/unique_count.cu]: Not familiar. IDEs sometimes add headers automatically?

**Review by ttnghia — APPROVED** [review]

File: cpp/src/stream_compaction/distinct.cu
Comment (davidwendt) [inline cpp/src/stream_compaction/distinct.cu:28]:
```suggestion
#include <rmm/mr/device/per_device_resource.hpp>

```
(adds an extra blank line between the rmm and thrust include sections)

File: cpp/tests/copying/detail_gather_tests.cu
Comment (davidwendt) [inline cpp/tests/copying/detail_gather_tests.cu:62]: This is an interesting change. 
The intention of this source file seems to be wanting to test `cudf::detail::gather` functions -- specifically the iterator interfaces as per this comment from above:
https://github.com/rapidsai/cudf/pull/12888/files#diff-88fa573b86cc7871ca7ce24babaa1651cb41a2c76de8fdab1925c21dae450fc4R44-R45
Perhaps these tests are really not needed and this source file itself is not necessary at all.

Comment (vyasr) [reply cpp/tests/copying/detail_gather_tests.cu:62]: Hmm yeah that is a good point. This file had two invocations of `detail::gather` and it looks like the other one was properly passing the stream through. I think the minimal change for this PR would be to just pass the stream, then to consider removing in a separate PR, WDYT? I made the change that I did thinking that it would be best to move away from using detail APIs in test altogether, but I also want to limit scope creep.

Comment (davidwendt) [reply cpp/tests/copying/detail_gather_tests.cu:62]: We can evaluate removing these tests in a separate PR.
But then I think you should change this back to calling the `detail` API right?

Comment (vyasr) [reply cpp/tests/copying/detail_gather_tests.cu:62]: Yup was waiting for your approval on the plan before pushing the changes. I'll update it now.

File: cpp/benchmarks/common/generate_input.cu
Comment (davidwendt) [inline cpp/benchmarks/common/generate_input.cu:35]:
```suggestion
#include <rmm/device_uvector.hpp>
#include <rmm/mr/device/per_device_resource.hpp>

```

File: cpp/src/rolling/detail/lead_lag_nested.cuh
Comment (davidwendt) [inline cpp/src/rolling/detail/lead_lag_nested.cuh:30]:
```suggestion
#include <rmm/exec_policy.hpp>
#include <rmm/mr/device/per_device_resource.hpp>

```

File: cpp/tests/copying/gather_str_tests.cpp
Comment (davidwendt) [inline cpp/tests/copying/gather_str_tests.cpp:27]:
```suggestion

#include <rmm/mr/device/per_device_resource.hpp>
```

**Review by davidwendt — APPROVED** [review]

**Pull request merged**
